pkg/upgrade/upgrademanager/upgrademanager_test: TestMigrationFailure failed

pkg/upgrade/upgrademanager/upgrademanager_test.TestMigrationFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10875121?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10875121?buildTab=artifacts#/) on master @ [87cd65352a9903c292d6b6a2e9856cc1f57bb267](https://github.com/cockroachdb/cockroach/commits/87cd65352a9903c292d6b6a2e9856cc1f57bb267):


```
=== RUN   TestMigrationFailure
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure2966434211
    test_log_scope.go:81: use -show-logs to present logs inline
    manager_external_test.go:687: test will fail at version: 1000022.2-62
    manager_external_test.go:767: 
        	Error Trace:	pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:767
        	            				pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:784
        	Error:      	Not equal: 
        	            	expected: roachpb.Version{Major:1000022, Minor:2, Patch:0, Internal:61}
        	            	actual  : roachpb.Version{Major:1000022, Minor:2, Patch:0, Internal:60}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -4,3 +4,3 @@
        	            	  Patch: (int32) 0,
        	            	- Internal: (int32) 61
        	            	+ Internal: (int32) 60
        	            	 }
        	Test:       	TestMigrationFailure
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure2966434211
--- FAIL: TestMigrationFailure (28.02s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-29654

pkg/upgrade/upgrademanager/upgrademanager_test.TestMigrationFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10882562?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10882562?buildTab=artifacts#/) on master @ [6a6ebcb950c050652ad8cded716dcd242e3a729c](https://github.com/cockroachdb/cockroach/commits/6a6ebcb950c050652ad8cded716dcd242e3a729c):


```
=== RUN   TestMigrationFailure
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure1447009331
    test_log_scope.go:81: use -show-logs to present logs inline
    manager_external_test.go:687: test will fail at version: 1000022.2-90
    manager_external_test.go:789: 
        	Error Trace:	pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:789
        	Error:      	Received unexpected error:
        	            	pq: verifying precondition for version 1000023.1-2: query-invalid_objects: crdb-internal-jobs-table: system-jobs-scan: column "job_type" does not exist
        	Test:       	TestMigrationFailure
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure1447009331
--- FAIL: TestMigrationFailure (26.87s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/upgrade/upgrademanager/upgrademanager_test.TestMigrationFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10886888?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10886888?buildTab=artifacts#/) on master @ [c21cfc2823728fde7fa6aecc9506a1afa1f83191](https://github.com/cockroachdb/cockroach/commits/c21cfc2823728fde7fa6aecc9506a1afa1f83191):


```
=== RUN   TestMigrationFailure
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure2048388363
    test_log_scope.go:81: use -show-logs to present logs inline
    manager_external_test.go:687: test will fail at version: 1000022.2-60
    manager_external_test.go:767: 
        	Error Trace:	pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:767
        	            				pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:784
        	Error:      	Not equal: 
        	            	expected: roachpb.Version{Major:1000022, Minor:2, Patch:0, Internal:59}
        	            	actual  : roachpb.Version{Major:1000022, Minor:2, Patch:0, Internal:58}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -4,3 +4,3 @@
        	            	  Patch: (int32) 0,
        	            	- Internal: (int32) 59
        	            	+ Internal: (int32) 58
        	            	 }
        	Test:       	TestMigrationFailure
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure2048388363
--- FAIL: TestMigrationFailure (26.93s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/upgrade/upgrademanager/upgrademanager_test.TestMigrationFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10887352?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10887352?buildTab=artifacts#/) on master @ [2a70922c5ce3d30373adfc52a46f869f9685497c](https://github.com/cockroachdb/cockroach/commits/2a70922c5ce3d30373adfc52a46f869f9685497c):


```
=== RUN   TestMigrationFailure
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure1830741605
    test_log_scope.go:81: use -show-logs to present logs inline
    manager_external_test.go:687: test will fail at version: 1000022.2-78
    manager_external_test.go:789: 
        	Error Trace:	pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:789
        	Error:      	Received unexpected error:
        	            	pq: verifying precondition for version 1000023.1-2: query-invalid_objects: crdb-internal-jobs-table: system-jobs-scan: column "job_type" does not exist
        	Test:       	TestMigrationFailure
    panic.go:522: -- test log scope end --
--- FAIL: TestMigrationFailure (27.54s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/upgrade/upgrademanager/upgrademanager_test.TestMigrationFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10888022?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10888022?buildTab=artifacts#/) on master @ [7339e0f900eba75db210874612d6e5aa561c854a](https://github.com/cockroachdb/cockroach/commits/7339e0f900eba75db210874612d6e5aa561c854a):


```
=== RUN   TestMigrationFailure
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure732820411
    test_log_scope.go:81: use -show-logs to present logs inline
    manager_external_test.go:687: test will fail at version: 1000022.2-78
    manager_external_test.go:789: 
        	Error Trace:	pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:789
        	Error:      	Received unexpected error:
        	            	pq: verifying precondition for version 1000023.1-2: query-invalid_objects: crdb-internal-jobs-table: system-jobs-scan: column "job_type" does not exist
        	Test:       	TestMigrationFailure
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure732820411
--- FAIL: TestMigrationFailure (27.07s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/upgrade/upgrademanager/upgrademanager_test.TestMigrationFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10888608?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10888608?buildTab=artifacts#/) on master @ [269b9e31c4befe53181340503663ae6e3e8ed07e](https://github.com/cockroachdb/cockroach/commits/269b9e31c4befe53181340503663ae6e3e8ed07e):


```
=== RUN   TestMigrationFailure
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure3478277740
    test_log_scope.go:81: use -show-logs to present logs inline
    manager_external_test.go:687: test will fail at version: 1000022.2-86
    manager_external_test.go:767: 
        	Error Trace:	pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:767
        	            				pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:784
        	Error:      	Not equal: 
        	            	expected: roachpb.Version{Major:1000022, Minor:2, Patch:0, Internal:85}
        	            	actual  : roachpb.Version{Major:1000022, Minor:2, Patch:0, Internal:84}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -4,3 +4,3 @@
        	            	  Patch: (int32) 0,
        	            	- Internal: (int32) 85
        	            	+ Internal: (int32) 84
        	            	 }
        	Test:       	TestMigrationFailure
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure3478277740
--- FAIL: TestMigrationFailure (32.64s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

pkg/upgrade/upgrademanager/upgrademanager_test.TestMigrationFailure [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10891690?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Ci_TestsAwsLinuxArm64_UnitTests/10891690?buildTab=artifacts#/) on master @ [2627dc50694837984d7bf7d6c6f25e294ae36039](https://github.com/cockroachdb/cockroach/commits/2627dc50694837984d7bf7d6c6f25e294ae36039):


```
=== RUN   TestMigrationFailure
    test_log_scope.go:167: test logs captured to: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure4157702422
    test_log_scope.go:81: use -show-logs to present logs inline
    manager_external_test.go:687: test will fail at version: 1000022.2-56
    manager_external_test.go:767: 
        	Error Trace:	pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:767
        	            				pkg/upgrade/upgrademanager/upgrademanager_test_test/pkg/upgrade/upgrademanager/manager_external_test.go:784
        	Error:      	Not equal: 
        	            	expected: roachpb.Version{Major:1000022, Minor:2, Patch:0, Internal:55}
        	            	actual  : roachpb.Version{Major:1000022, Minor:2, Patch:0, Internal:54}
        	            	
        	            	Diff:
        	            	--- Expected
        	            	+++ Actual
        	            	@@ -4,3 +4,3 @@
        	            	  Patch: (int32) 0,
        	            	- Internal: (int32) 55
        	            	+ Internal: (int32) 54
        	            	 }
        	Test:       	TestMigrationFailure
    panic.go:522: -- test log scope end --
test logs left over in: /artifacts/tmp/_tmp/be5347fadc06714556335e3ce75bfe99/logTestMigrationFailure4157702422
--- FAIL: TestMigrationFailure (28.25s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestMigrationFailure.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/jobs/utils.go
+++ b/pkg/jobs/utils.go
@@ -156,16 +156,24 @@ func JobExists(
 	return row != nil, nil
 }
 
-// IsJobTypeColumnDoesNotExistError returns true if the error is of the form
+// isJobTypeColumnDoesNotExistError returns true if the error is of the form
 // `column "job_type" does not exist`.
 func isJobTypeColumnDoesNotExistError(err error) bool {
 	return pgerror.GetPGCode(err) == pgcode.UndefinedColumn &&
 		strings.Contains(err.Error(), "column \"job_type\" does not exist")
 }
 
+// isJobInfoTableDoesNotExistError returns true if the error is of the form
+// `related "job_info" does not exist`.
+func isJobInfoTableDoesNotExistError(err error) bool {
+	return pgerror.GetPGCode(err) == pgcode.UndefinedTable &&
+		strings.Contains(err.Error(), "relation \"system.job_info\" does not exist")
+}
+
 // MaybeGenerateForcedRetryableError returns a
 // TransactionRetryWithProtoRefreshError that will cause the txn to be retried
-// if the error is because of an undefined job_type column.
+// if the error is because of an undefined job_type column or missing job_info
+// table.
 //
 // In https://github.com/cockroachdb/cockroach/issues/106762 we noticed that if
 // a query is executed with an AS OF SYSTEM TIME clause that picks a transaction
@@ -173,15 +181,19 @@ func isJobTypeColumnDoesNotExistError(err error) bool {
 // infrastructure will attempt to query the job_type column even though it
 // doesn't exist at the transaction's timestamp.
 //
-// As a short term fix, when we encounter an `UndefinedColumn` error we
-// generate a synthetic retryable error so that the txn is pushed to a
-// higher timestamp at which the upgrade will have completed and the
-// `job_type` column will be visible. The longer term fix is being tracked
-// in https://github.com/cockroachdb/cockroach/issues/106764.
+// As a short term fix, when we encounter an `UndefinedTable` or
+// `UndefinedColumn` error we generate a synthetic retryable error so that the
+// txn is pushed to a higher timestamp at which the upgrade will have completed
+// and the table/column will be visible. The longer term fix is being tracked in
+// https://github.com/cockroachdb/cockroach/issues/106764.
 func MaybeGenerateForcedRetryableError(ctx context.Context, txn *kv.Txn, err error) error {
 	if err != nil && isJobTypeColumnDoesNotExistError(err) {
 		return txn.GenerateForcedRetryableError(ctx, "synthetic error "+
 			"to push timestamp to after the `job_type` upgrade has run")
+	}
+	if err != nil && isJobInfoTableDoesNotExistError(err) {
+		return txn.GenerateForcedRetryableError(ctx, "synthetic error "+
+			"to push timestamp to after the `job_info` upgrade has run")
 	}
 	return err
 }

--- a/pkg/upgrade/upgrademanager/BUILD.bazel
+++ b/pkg/upgrade/upgrademanager/BUILD.bazel
@@ -68,7 +68,6 @@ go_test(
         "//pkg/sql/sem/eval",
         "//pkg/testutils",
         "//pkg/testutils/serverutils",
-        "//pkg/testutils/skip",
         "//pkg/testutils/sqlutils",
         "//pkg/testutils/testcluster",
         "//pkg/upgrade",

--- a/pkg/upgrade/upgrademanager/manager_external_test.go
+++ b/pkg/upgrade/upgrademanager/manager_external_test.go
@@ -38,7 +38,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
 	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/upgrade"
@@ -682,8 +681,6 @@ func TestPrecondition(t *testing.T) {
 func TestMigrationFailure(t *testing.T) {
 	defer leaktest.AfterTest(t)()
 	defer log.Scope(t).Close(t)
-
-	skip.WithIssue(t, 106648)
 
 	ctx := context.Background()
 

